### ISSUE TITLE
*Title Edit* File mapping reading locked file = Crash

### ISSUE THREAD

**Issue by Tigzy** [VirusTotal/yara#979]
Here's a snapshot of a crash dump I have. It seems at some point the yr_scanner_scan_mem function iterates index for block_data a little bit too much. The array position is pointing in a unavailable memory address and the program crashes. Any direction for me to look at? Do you want the dump and symbols?

![image](https://user-images.githubusercontent.com/9531438/48117934-f8009a80-e26a-11e8-8b40-cf80cc7c1f40.png)

**Comment by Tigzy** [comment]
I cannot reproduce the crash myself, it's a dump uploaded from a client.

**Comment by Tigzy** [comment]
Just to add a little bit of context, and I think this might be useful, at the moment of the crash I have 2 other worker threads: 
- One iterating strings (yr_rule_strings_foreach)
- Another one blocked the EXACT SAME line as the crashing one, so I'm pretty sure it's a thread-safety issue. Another weird thing, the i variable has the same value (even though the block_data (mapped file pointer) is different:

0x000001724f8f0000 + x2600
0x000001724f9e0000 + x2600

![image](https://user-images.githubusercontent.com/9531438/48118529-d4d6ea80-e26c-11e8-9b38-a7f8bf8c26f2.png)

**Comment by Tigzy** [comment]
Up? I've dozens of dumps with the same pattern. 
i grows up until a value where the memory cannot be read. I suspect the boundaries in the block struct are wrong, or badly calculated.

**Comment by Joseph-Tortorelli** [comment]
Wait, so you are getting multiple dumps with the same pattern? Something really weird has recently been going on. People just found malware in some big repositories. I wonder if something is preventing YARA from working. What if someone was able to break YARA and therefore, stop their own malware's signature from uploading? Maybe I am overthinking this but so much weird stuff has just been going on

**Comment by Tigzy** [comment]
> Wait, so you are getting multiple dumps with the same pattern? Like from different people/clients? Something really weird has recently been going on. People just found malware in some big repositories. I wonder if something is preventing YARA from working. Maybe I am overthinking this but so much weird stuff has just been going on

Na I don't think so. It looks really typical of a thread-safety issue. I think there's some shared variable being corrupted and not guarded because the the crash context is very similar everytime, with 2 threads entering the same code area at the same time

**Comment by Tigzy** [comment]
Adding some more context: 
I see the callback is called just after that problematic block, with RULE* as parameter. The rule pointer appears to be some block from the shared RULES* pointer, and is not guarded. 

It if were read only that would be fine but I can see is has a list of strings (OK) having a list of matches (NOK). Those matches are likely to change at every different scanned object right? Should I guard the whole matches read while in the callback?

**Comment by Tigzy** [comment]
TLDR; Is RULES* pointer sent in callback is thread safe ?

**Comment by Tigzy** [comment]
I think I found the issue, this isn't about thread safety after all. 
The file Yara is trying to map at the moment of the crash is driver-protected and cannot be read. 
I'll try to get the error code returned by the file mapping API.

![image](https://user-images.githubusercontent.com/9531438/49083694-a89c0180-f24d-11e8-9176-3ccb1aeda08f.png)

**Comment by Tigzy** [comment]
ReadFile fails with DEVICE DATA ERROR. Since we are using a file mapping here, how can we avoid this? Can we check buffer access before reading?

![image](https://user-images.githubusercontent.com/9531438/49084105-d3d32080-f24e-11e8-87de-62df7993ae63.png)

**Comment by Tigzy** [comment]
So the idea is that NVIDIA driver is preventing yara from reading certain blocks inside the file. How can this be catched?

**Comment by Tigzy** [comment]
OK so it's obvious the culprit line is: 
` index = block_data[i++] + 1;` (scanner.c / line 86)

When I read the documentation about file mapping read, it says: 
https://msdn.microsoft.com/en-us/library/windows/desktop/aa366761(v=vs.85).aspx
"To guard against EXCEPTION_IN_PAGE_ERROR exceptions, use structured exception handling to protect any code that writes to or reads from a memory mapped view of a file other than the page file. For more information, see Reading and Writing From a File View."

Here: https://docs.microsoft.com/fr-fr/windows/desktop/Memory/reading-and-writing-from-a-file-view
This is how we should read from the mapped file. 
```
DWORD dwLength;

  __try
  {
    dwLength = *((LPDWORD) lpMapAddress);
  }
  __except(GetExceptionCode()==EXCEPTION_IN_PAGE_ERROR ?
    EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH)
  {
    // Failed to read from the view.
  }
```

**Comment by Tigzy** [comment]
I'll build a modified Yara version to test it on crashing machine, if it works I'll submit a PR.

**Comment by Tigzy** [comment]
I can CONFIRM the guard fixes the issue. 
I've made a PR: https://github.com/VirusTotal/yara/pull/991

**Comment by Tigzy** [comment]
I realize I made this PR like a Windows barbarian 🗡 Feel free to adapt for linux with preprocessor directives...
